drivers/gpio/gpio_npcx.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_gpio
+
+#include <kernel.h>
+#include <device.h>
+#include <drivers/gpio.h>
+#include <soc.h>
+
+#include "gpio_utils.h"
+#include "soc_gpio.h"
+#include "soc_miwu.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_npcx, LOG_LEVEL_ERR);
+
+/* GPIO module instances declarations */
+static struct device *gpio_devs[];
+static int gpio_devs_count;
+
+/* Driver config */
+struct gpio_npcx_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	/* GPIO controller base address */
+	uint32_t base;
```

Comment:
I have tried it before. It will cause a warning during code style checking. Please see below.

_-:9: WARNING:NEW_TYPEDEFS: do not add new typedefs
#9: FILE: drivers/gpio/gpio_npcx.c:55:
+typedef struct gpio_reg gpio_reg_t;

- total: 0 errors, 1 warnings, 7 lines checked_

Reply:
I think the static cast in the initialization structure make the most sense because you are trying to format the data in the best representation there. I would rather see the cast once in initialization rather than having the cast every time we call it. Also specifying the datatype here in the structure makes it easier to know what type of data to expect.

Reply:
really would prefer you use `uintptr_t` instead of `uint32_t` for linear addresses treated as integral values

Reply:
Ok. I will add one more commit here to use uintptr_t instead of uint32_t including the previous PR.

Reply:
Done.